DBZ-1079 Add a sortable ID to each message

addisonj opened the pull request:
This commit adds a new optional ID that can be used to order the
messages.

It introduces a new configuration option, `ADD_ORDERED_ID` that appends
a new field `ulid` which is a unique, lexigraphically sortable id.

This is tested in postgres, but should work in mysql as well.

jpechane commented:
@addisonj Hi, I like the PR - to provide a totally orderd over over all messages. I'll refer to @gunnarmorling for the final ruesult but so far I have few notes
1. IMHO, this is very useful and should be limited to embedded mode but should be supported by connectors even in Kafka Connect
2. The value should be writtent into `source` block in the envelope, not to the envelope itself
3. The value shouls be also written to offsets and read from offsets when the connector is started. RIght now you'd reset the counter between runs which is not desirable. Ta last value should be recovered from offsets and the counter should be refreshed from it.
4. I am not 100 % sure about introducing `sulky` as dependency. Also I'd like to see plain `long` based counter to be available out-of-the-box

gunnarmorling commented:
Interesting suggestion; thanks a lot, @addisonj!

> IMHO, this is very useful and should be limited to embedded mode but should be supported by connectors even in Kafka Connect

@jpechane, is there a "not" missing: "should *not* be limited to embedded"?

> The value should be writtent into source block in the envelope, not to the envelope itself

+1

> The value shouls be also written to offsets and read from offsets when the connector is started. 

Yes, so that's the biggest question I have around this. Also with storing it into the offsets it could happen that the same sequence value is used another time after a connector restart. I'd still like to better understand the use case to see whether that'd be an acceptable limitition or not.

> Also I'd like to see plain long based counter to be available out-of-the-box

+1, something based on `AtomicLong` could be useful? It seems that Sulky would prevent the creation of more than 1,000 messages per second?

addisonj commented:
Yeah, I debated between the envelope and the SourceInfo, the biggest reason I went for the Envelope is to make it configurable (which was going to be trickier to do with SourceInfo) as well as the fact that each connector uses it's own SourceInfo, so sharing the code was more problematic, but I actually didn't notice the "AbstractSourceInfo" so perhaps it makes sense to do a bit of a refactor to move it there.

As far as the choice of a long vs the ulid, the reason for the ulid is a few fold:
1. it doesn't need to be persisted while still giving quite strong guarantees (will touch on in a minute) and is also independent of state, so for example, if I am not doing any snapshots and I lose my debezium state, I still get a correct ordering without the long resetting. This seems more flexible that a simple counter. 
2. With a counter, you are still going to see the counter "roll back" under failure as events are delivered more than once. This *could* be desirable, but personally I would rather deal with duplicates using something like txid/table/pk as a key to drop dupes and still have my ordering not result in duplicates being grouped together. In other words, I think you want this ID to be advancing even under failure and a true ordering.
3. As far as the concerns over guarantees, this should be very safe, the two ways in which this could be broken are stated below:
- If the process restarts on the same server in the same millisecond, you could potentially have ids go backwards, however, that seems quite unlikely just due to JVM startup time, etc
- If the process is moved to a new server with a different clock skew, it could go backwards, but since that moving is likely to take a fairly long period of time, so long as the user has NTP, it is unlikely.
That said, it is possible to persist it into the state and ensure that it is greater than what we last new about, but that doesn't buy us much more of a guarantee for keeping it globally ordered (see 2), because we aren't guaranteed that the ID persisted to state is in fact the most recent id generated.

As far as limits on numbers of IDs generated by ULID, it is a bit more complex than that. For each new millisecond, it choose a random number with 80 bits of randomness and then increments from there, so it is sorta like `Long.MAX_VALUE - Random.nextLong()` (but actually bigger than a long) unique IDs per millisecond, each millisecond. So yes, it is *possible* you could get very unlucky and only have a small number of IDs per ms, but it is exceedingly unlikely (like once every few years if you generate a million ULIDs a ms). That said, it is pretty easy to protect against, which is why you see that sleep to wait for the next millisecond.

Perhaps the correct answer is to make both options available to the user, which is fairly easily done.

addisonj commented:
@gunnarmorling @jpechane 

I reworked this heavily, but before I go too much deeper, I want your eyes.

An overview of my changes:
- Moved the id generation into AbstractSourceInfo which now requires the idBuilder be passed. This is pretty simple in the the code but does add a bit of cruft to tests.
- Made it so multiple different implementations can be used
- The SourceInfo is now responsible responsible to call `buildNextId`, which is different for each connector.

The biggest question I have:
Currently, each SourceInfo implementation calls `buildNextId` in the methods that track new state. This seems safe, but not super consistent and potentially if multiple messages reuse an offset (postgres doesn't do this, but not sure about mysql and mongo) you wouldn't get the updated offset and share an ID. The alternative seems to be to have the worker be responsible for generating the next id and pass it to source info, but that seems like it would lead to poor code sharing and more cruft. If there is concern about the current approach, what I sorta thought would work is a new method on SourceInfo that *must* be called for each message.

What is left:
Some tests, primarily tests for each SourceInfo implementation and an integration test for each.

gunnarmorling commented:
> In other words, I think you want this ID to be advancing even under failure and a true ordering.

This is the part where I'm still struggling to wrap my head around this. So after a connector restart it could happen an event will be emitted which has been emitted before. But it will have a different ulid this time. I think it'd help to see an actual example that shows, how that new sequence value would exactly be used.

gunnarmorling commented on debezium-core/src/main/java/io/debezium/util/OrderedIdBuilder.java:
```diff
@@ -0,0 +1,47 @@
+
+import org.apache.kafka.connect.data.Schema;
+
+public abstract class OrderedIdBuilder {
```
How about this contract:

```
public abstract class OrderedIdBuilder {
    public void addOrderedId(Struct sourceInfo);
    public void restoreState(String lastState);
    public OrderedIdBuilder clone(); // not quite sure on that one yet
}
```

That'd move most code into implementations of this contract themselves (e.g. the no-op handling would be fully encapsulated), simplifying the code in the connectors (whose `SourceInfo#struct()` method just would have to call `addOrderedId()` whenever invoked).

addisonj replied:
As far as I can tell, all the implementations of SourceInfo look more like this:

1. update the position in the log (and how I have implemented it, bump the order_id)
2. build the current position as a `Map<String, Object>`  to store in the offsetStore
3. build the struct that is actually written out in a message

So I think there are two issues with that API:
1. The struct and the map are both required, especially for the counter,  so we would at least need a method to get the current state so we can also generate the map that is written to the offsetStore.
2. Because the calls happen in the order described above, if you don't get the next id until you actually build the struct, then the map you build will be *before* you generate the new ID, meaning there will be an off by one error. 

I think a simpler API could exist, but it seems like it would need to change with SourceInfo API to be more controlled for that to work.

gunnarmorling commented on debezium-core/src/main/java/io/debezium/util/UlidIdBuilder.java:
```diff
@@ -0,0 +1,67 @@
+            // and ignore any interrupts, as we can just try
+            // again on the next loop
+            try {
+                Thread.sleep(1);
```
Ah, I think I misunderstood this part before. It will only sleep in case there _is_ a collision already, right?

addisonj replied:
correct

addisonj replied:
Also, just as a reminder that should be _exceedingly_ rare, like it may never happen in reality, but we handle it anyways

addisonj commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/SourceInfo.java:
```diff
@@ -241,6 +248,8 @@ public Struct lastOffsetStruct(String replicaSetName, CollectionId collectionId)
     public Struct offsetStructForEvent(String replicaSetName, Document oplogEvent) {
         Position position = INITIAL_POSITION;
+        OrderedIdBuilder idBuilder = idBuildersByReplicaSetName.getOrDefault(replicaSetName, idBuilder().clone());
+        idBuilder.buildNextId();
```
From my understanding and reading the code, in the case of mongodb with sharding, there will be multiple replica sets being tracked independently. This means that we essentially have multiple streams, and ideally, each with their own independent order_i. So we need to create a new OrderedIdBuilder for each replica set. 

Perhaps I misunderstood that though?

addisonj commented on debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsStreamProducerIT.java:
```diff
@@ -1106,6 +1107,43 @@ public void shouldReceiveChangesForTableWithoutPrimaryKey() throws Exception{
         assertRecordSchemaAndValues(expectedAfter, deletedRecord, Envelope.FieldName.AFTER);
     }
 
+    /*
```
that test was for the old implementation, I just wanted to get feedback on the direction of this before going further with tests and docs/cleanup

addisonj commented on debezium-core/src/main/java/io/debezium/util/OrderedIdBuilder.java:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * clones this id builder along with state.
+     *
+     * This is primarily used for connectors to track multiple
```
See the above comment, but yes, essentially for where multiple streams exists and allows them to be incremented independently.

addisonj commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SourceInfo.java:
```diff
@@ -380,6 +380,7 @@ public void startNextTransaction() {
      * Capture that we're starting a new event.
      */
     public void completeEvent() {
+        getNextId();
```
not as currently done, see the below comment in regards to your changed API contract for a more complete answer, but essentially, the way SourceInfo is interacted with is always to first update the state and then build representations of that state for persisting to the offsetStore as well as the message itself.

In other words, we don't do anything with it, because we don't have anywhere to store it yet, that instead happens once we call `struct(TableId)` and `offset()`

addisonj commented:
That @gunnarmorling for the feedback, hopefully my answers make sense. But to your first question about MySql, yeah, that would be a bit more complicated, but not totally unworkable as currently designed. As mentioned above, this already sorta happens with the mongodb connector, which (AFAICT) uses a different stream for each replica set. A similar technique could be used here, where each binlog reader would have it's own ID. When the two readers eventually do converge, you could take the greater of both ids. I haven't looked much at the db-175 branch yet to see the work entailed, but it would probably take some effort.

Anyways, let me know what you think, I don't have a *ton* of time to spend on this, but am happy to try and get this finished up!

addisonj commented:
@gunnarmorling small bump on this, curious if you have any more feedback in response to my answers.

I am happy to tweak this a bit more, but not quite sure what the better option is without a bigger refactor to SourceInfo.

gunnarmorling commented:
Hey @addisonj, thanks for the bump, there's just so many things going on... . So I think it'd help to take a step back and get some more clarity about the goal we want to achieve here. We're heads down to implementation questions (and I brought them up, so mea culpa ;), but actually it's still not totally clear to me _what_ the purpose of this is. Could you check out my last comment on [DBZ-1079](https://issues.jboss.org/browse/DBZ-1079) and provide an example of how you're planning to use this? I think that'd be very helpful to get clarity about the _what_ before deciding on the _how_. Thanks!

addisonj commented:
First... why I want an id that gives me an ordering:

My use case for debezium is doing realtime ETLs using a stream processing engine (specifically, Apache Flink). Some of these ETLs will be doing joins across tables and even across schemas, so as it stands, debezium doesn't give me a way to know the order of changes between tables by just looking at the data. With a parallel stream processor like Flink, I can't directly control that all the partitions will be consumed evenly, which means that doing a join between two streams could potentially result in an incorrect join and an incorrect answer. However, this is made even more difficult by the fact that I am using Kinesis, which has such small partitions that for some tables with lots of writes, I will need to split the writes for that table across multiple partitions, which completely throws out being able to use the log as a source of ordering. 

My way out of this though is that with a stream processor like Flink and an ID in the data that provides order, I can easily re-order the changes over some domain (table, schema, database, etc) and then process the results. 

One more thing I mention, I really don't need this to be in debezium core because of the fact I am using EmbeddedEngine, however, we may potentially switch to Kafka so we would prefer to have as much as possible in Debezium, as well as it seems like any serious use of debezium with postgres is going to run into some ordering issues, if only when doing joins across tables.

Yes, so I agree this has gotten quite big...While I think this *potentially* might be a useful feature for all connectors, it may not be as critical for them, as they may have existing fields that allow for good ordering.

However, with postgres, that is definitely not the case. The LSN can't be used, as it is the LSN at each row change, and since those can be interleaved with multiple concurrent transactions, they can be all over the place.  While there is a timestamp (either the commit timestamp or the debezium timestamp) this obviously has issues with collisions and depending on if you use decoderbufs or wal2json, you aren't getting commit timestamp, but just a process generate timestamp.

So AFAICT, there isn't a way to get order out of debezium with Postgres at least, but, it seems like it might be a bit of a stretch to implement this everywhere just yet. 

What I might propose is this: Remove all of the changes in the mysql, and mongo connectors, but leave the `OrderedIdBuilder` and subclasses in debezium-core. Additionally, remove any configuration for it out of `CommonConnectorConfig` and remove the changes for `AbstractSourceInfo` and instead implement this just for postgres.

That would allow for at least some reuse of this functionality in the future, but would vastly reduce the complexity. Once again, I can still do this myself due to using EmbeddedEngine, so if it still feels like something you still aren't sure about, it won't impact me if it is abandoned.

gunnarmorling commented:
Hey @addisonj, sorry for the silence on that one, was heads-down into other stuff. Still need to fully process your answer, too. That said you mention moving to Kafka proper at some point, would be using an SMT (single message transform) for adding the id a way out in that case for you?

addisonj commented:
I am not so familiar with Kafka Connect, but the biggest problem would be that if the transform happens after records are written to a topic, you would loose some ordering already (as it could be split across multiple partitions), if you can do an SMT before the data is written to a topic, that would work.

gunnarmorling commented:
> but the biggest problem would be that if the transform happens after records are written to a topic

You can apply SMTs to source connectors (such as Debezium) or sink connectors. In case of the former, they process the message *before* it's written to a topic (which allows an SMT to re-route to a different topic than originally intended or even drop a message). So the ordering there would be the exact same order of events as retrieved from the DB log.

addisonj commented:
Going to close this, while there is a minor loss in doing it outside of debezium core (in that you can't easily do integer IDs easily) this seems like it adds a lot of complexity. Eventually, it seems like this might be a nice feature in debezium, but it isn't worth the complexity right now.
